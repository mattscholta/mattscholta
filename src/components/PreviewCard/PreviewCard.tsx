import * as React from 'react';
import classnames from 'classnames';
import Link from 'next/link';

import styles from './styles.module.css';

export interface PreviewCardProps {
  className?: string;
  description: string;
  image: string;
  slug: string;
  title: string;
}

/**
 * @name PreviewCard
 * @external https://www.w3schools.com/howto/howto_css_flip_box.asp
 * @description Preview cards that use a CSS flip to reveal a bit more data
 */
export const PreviewCard = (props: PreviewCardProps) => {
  const { className, description, image, slug, title } = props;

  // Styles
  const css = classnames(styles.card, className);

  return (
    <div className={css}>
      <div className={styles.inner}>
        <div className={styles.front}>
          <img
            alt={title}
            height={200}
            loading="eager"
            src={image}
            title={title}
            width={300}
          />
        </div>
        <div className={styles.back}>
          <Link href={`/portfolio/${slug}`}>
            <a className="u-flex u-flex-column">
              <h2 className="ui-heading medium">{title}</h2>
              <p className="u-flex-grow --u-overflow-ellipsis">{description}</p>
              <div className={styles.link}>Read More &#187;</div>
            </a>
          </Link>
        </div>
      </div>
    </div>
  );
};
