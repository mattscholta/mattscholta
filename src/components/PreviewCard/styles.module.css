.card {
  background-color: transparent;
  cursor: pointer;
  height: 200px;
  margin-bottom: var(--grid-4x);
  perspective: 1000px; /* Remove this if you don't want the 3D effect */
  width: 300px;
}

.inner {
  height: 100%;
  position: relative;
  text-align: center;
  transform-style: preserve-3d;
  transition: transform 360ms ease-in;
  width: 100%;
}

.card:hover .inner {
  transform: rotateX(-180deg);
}

.back,
.front {
  -webkit-backface-visibility: hidden; /* Safari */
  backface-visibility: hidden;
  height: 100%;
  position: absolute;
  width: 100%;
}

.back {
  background-color: var(--spot-white);
  border-radius: var(--grid-1x);
  color: var(--spot-grey-7);
  padding: 0 var(--grid-2x);
  transform: rotateX(-180deg);
}

.front {
  background-color: var(--spot-white);
}

.link {
  color: var(--spot-primary);
  font-size: var(--font-14);
  font-weight: var(--font-weight-medium);
  margin-bottom: var(--grid-2xs);
  text-align: right;
}

.back a {
  align-items: stretch;
  height: 100%;
  justify-content: center;
}

.back h2 {
  font-weight: var(--font-weight-bold);
  margin-top: 24px;
  text-align: left;
}

.card p {
  font-size: 0.9em;
  line-height: 1.2;
  margin-top: var(--grid-2x);
  text-align: left;
}

.card a,
.card a:hover {
  color: var(--spot-grey-7);
  text-decoration: none;
}
