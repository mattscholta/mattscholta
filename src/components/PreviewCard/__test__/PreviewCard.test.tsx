import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';

import { PreviewCard, PreviewCardProps } from '../PreviewCard';

describe('PreviewCard', () => {
  let props: PreviewCardProps;
  let wrapper: RenderResult;

  const createWrapper = () => <PreviewCard {...props} />;

  beforeEach(() => {
    props = {
      data: {
        description: 'test-description',
        image: 'test-image',
        slug: 'test-slug',
        title: 'test-title'
      }
    };

    wrapper = render(createWrapper());
  });

  it('matches snapshot', () => {
    expect(wrapper.container.childNodes).toMatchSnapshot();
  });
});
