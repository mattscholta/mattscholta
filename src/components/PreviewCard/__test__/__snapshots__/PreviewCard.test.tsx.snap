// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewCard matches snapshot 1`] = `
NodeList [
  <div
    class="card"
  >
    <div
      class="inner"
    >
      <div
        class="front"
      >
        <img
          alt="test-title"
          height="200"
          loading="eager"
          src="test-image"
          title="test-title"
          width="300"
        />
      </div>
      <div
        class="back"
      >
        <a
          class="u-flex u-flex-column"
          href="/portfolio/test-slug"
        >
          <h2
            class="ui-heading medium"
          >
            test-title
          </h2>
          <p
            class="u-flex-grow --u-overflow-ellipsis"
          >
            test-description
          </p>
          <div
            class="link"
          >
            Read More Â»
          </div>
        </a>
      </div>
    </div>
  </div>,
]
`;
